loopy/statistics.py

```diff
@@ -0,0 +1,270 @@
+from __future__ import division
+from __future__ import absolute_import
+import six
+
+__copyright__ = "Copyright (C) 2015 James Stevens"
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+import numpy as np
+import loopy as lp
+import warnings
+from islpy import dim_type
+import islpy._isl as isl
+from pymbolic.mapper import CombineMapper
+
+
+class TypedPolyDict:
+
+    def __init__(self, i32=0, f32=0, f64=0):
+        self.poly_dict = {
+                            np.dtype(np.int32): i32,
+                            np.dtype(np.float32): f32,
+                            np.dtype(np.float64): f64}
+
+    def __add__(self, TPD):
+        return TypedPolyDict(
+            self[np.dtype(np.int32)]+TPD[np.dtype(np.int32)],
```

Comment:
The extra hoop through TPD.__setitem__ is not really necessary (and costs a little in performance)--just hit polydict directly.